lib/flakes.nix

```diff
@@ -19,4 +22,81 @@ rec {
     in
     self;
 
+  /*
+    Builds a map from <attr>=value to <attr>.<system>=value for each system,
+    except for the `hydraJobs` attribute, where it maps the inner attributes,
+    from hydraJobs.<attr>=value to hydraJobs.<attr>.<system>=value.
+  */
+  eachSystem = systems: f:
+    let
+
+      /*
+        Used to match Hydra's convention of how to define jobs. Basically transforms
+
+            hydraJobs = {
+              hello = <derivation>;
+              haskellPackages.aeson = <derivation>;
+            }
+
+        to
+
+            hydraJobs = {
+              hello.x86_64-linux = <derivation>;
+              haskellPackages.aeson.x86_64-linux = <derivation>;
+            }
+
+        if the given flake does `eachSystem [ "x86_64-linux" ] { ... }`.
+      */
+      pushDownSystem = system: merged:
+        mapAttrs
+          (name: value:
+            if ! (isAttrs value) then value
+            else if isDerivation value then (merged.${name} or {}) // { ${system} = value; }
+            else pushDownSystem system (merged.${name} or {}) value);
+
+      # Merge together the outputs for all systems.
+      op = attrs: system:
+        let
+          ret = f system;
+          op = attrs: key:
+            let
+              appendSystem = key: system: ret:
+                if key == "hydraJobs"
+                  then (pushDownSystem system (attrs.hydraJobs or {}) ret.hydraJobs)
+                  else { ${system} = ret.${key}; };
+            in attrs //
+              {
+                ${key} = (attrs.${key} or { }) // (appendSystem key system ret);
+              }
+          ;
+        in
+        foldl' op attrs (attrNames ret);
+    in
+    foldl' op { } systems
+  ;
+
+  /*
+    Returns the structure used by `nix app`
+
+    Example:
+      mkApp neovim
+      => { type = "app"; program = "/nix/store/.../bin/nvim"; }
+  */
+  mkApp = drv: {
+    type = "app";
+    program = exec drv;
+  };
+
+  /*
+    Returns the structure used by `nix app`
+
+    Example:
+      mkApp libnotify "notify-send"
+      => { type = "app"; program = "/nix/store/.../bin/notify-send"; }
+  */
+  mkApp' = drv: binName: {
```

Comment:
Wouldn't it be easier to combine those and fall back to the first argument of the second is not defined?

Reply:
I don't think it's possible to do it that way? The best I can think of in terms of merging is to swap arguments and check if the first argument is a derivation or a string, if its a string - return a function that accepts derivation

Reply:
Actually, another thought occurred - one argument that accepts either attribute set (same as flake-utils) or derivation. This would be a pretty clean implementation.

Anyways, let's wait and see what supreme nix'ers has to say about all this, not sure what are your thoughts on the matter but I think its time for reviewers to give in and merge this or something similar